### ISSUE TITLE
DISCUSSION :: Should we use maven-checkstyle to enforce Java coding style?

### ISSUE THREAD

**Issue by nickcmaynard** [swagger-api/swagger-codegen#3972]
##### Description
Our code is mostly formatted according to Google Java Style, but we can enforce this pretty easily using maven-checkstyle: https://maven.apache.org/plugins-archives/maven-checkstyle-plugin-2.16/

##### Swagger-codegen version

All.

**Comment by nickcmaynard** [comment]
Am happy to make this happen, if people think it's a good plan.

**Comment by wing328** [comment]
Sounds good to me.

**Comment by nickcmaynard** [comment]
So I have an outline plan for this, and it looks something like this:
1. Put in place the POM changes that execute checkstyle, but only WARN on changes - this is PR #4047.
2. We say "PRs don't increase the number of warnings" - ie. any changes don't fail checkstyle.
3. I'll submit a once-off PR that:
  1. Fixes all point-in-time checkstyle failures.
  2. Changes the POM checkstyle behaviour to ERROR (and fail the build).

What do people think?

**Comment by ePaul** [comment]
Hmm, is there some way to automatically fix the violations? I would like to not just annoy people, but give them an easy way to also fix the problems then.

**Comment by nickcmaynard** [comment]
I do have an uncrustify config that could go a long way to achieving that.  Will test it, use it, and submit it as part of stage 3.

**Comment by cbornet** [comment]
If we use google style, we can use [google-java-format](https://github.com/google/google-java-format) with `aosp`  option for 4-spaces indent

**Comment by nickcmaynard** [comment]
That could well be better...

**Comment by nickcmaynard** [comment]
@wing328 Now that #4047 is merged, we can update the PR template to say "check your code for checkstyle failures".  We can do that now or when (3) gets done (working on that), but I'd suggest we do it sooner rather than later.

**Comment by nickcmaynard** [comment]
Actually, I'm suspicious that we won't get to (3) against these strict Google styles - not without many man-hours.  A number of these checkstyle failures aren't automatically fixable - they include things like variable naming, etc.

I'll take a look at this, but we might be better served by moderating Google's style - some reasonable compromises.

**Comment by wing328** [comment]
Yup, agreed with you that it would take a lot of man-hours to complete (3). We'll leverage the community to work on it bit by bit.

**Comment by nickcmaynard** [comment]
#4214 takes care of the vast majority of the warnings once the formatter has been run once.  The remainder are not "formatting" issues - they're things like import ordering, enforcement of block-style `if` statements, etc.

I propose that once we merge #4214, a maintainer immediately runs `mvn formatter:format` on the codebase and commits the results of this.  Once that's done, we can hook the formatter into a standard mvn lifecycle phase - commented config already exists in `swagger-codegen/pom.xml`.
